### DISCUSSION TITLE
Multiple heads, downgrade single branch to branch point

### DISCUSSION THREAD

**Discussion by lafrech** [sqlalchemy/alembic#1480] [Usage Questions] [answered]
Working on two separate features concurrently, here's my migration tree:

```
rev1 -> rev2
     -> rev3
```

After upgrading to `heads`, `current` shows both `rev2 (head)` and `rev3 (head)`. Fine.

Now, say I'd like to revert `rev3` and keep `rev2`.

Downgrading to `rev1` downgrades both branches. Not what I want.

I've been reading https://alembic.sqlalchemy.org/en/latest/branches.html and https://github.com/sqlalchemy/alembic/issues/464.

I didn't use branch labels but since branches can also be referred to by rev name, I tried to downgrade `rev3@rev1` but it seems this downgrades both `rev2` and `rev3` branches. In fact, it seems anything before the `@` is ignored because whatever random string I put here doesn't seem to make a difference.

Is there something I'm misunderstanding?

Is there any way to achieve what I'm trying to?

If that matters, I'm using a custom command that calls the API as

```py
alembic.command.downgrade(config_path, "rev3@rev1")
```

**Comment by zzzeek** [comment, marked as answer]
the `rev3@rev1` approach works on this end, here's an example

```
[classic@framework alembic:main]$ PYTHONPATH=~/dev/sqlalchemy/lib/ python -m alembic current --verbose
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
Current revision(s) for postgresql://scott:***@localhost/test:
Rev: a815b8836ff0 (head)
Parent: 5f32c059b464
Path: /home/classic/dev/alembic/foo/versions/a815b8836ff0_rev3.py

    rev3
    
    Revision ID: a815b8836ff0
    Revises: 5f32c059b464
    Create Date: 2024-05-29 12:22:46.306333

Rev: 185d278c5837 (head)
Parent: 5f32c059b464
Path: /home/classic/dev/alembic/foo/versions/185d278c5837_rev2.py

    rev2
    
    Revision ID: 185d278c5837
    Revises: 5f32c059b464
    Create Date: 2024-05-29 12:22:04.837350

[classic@framework alembic:main]$ PYTHONPATH=~/dev/sqlalchemy/lib/ python -m alembic downgrade a815b8836ff0@5f32c059b464
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
INFO  [alembic.runtime.migration] Running downgrade a815b8836ff0 -> 5f32c059b464, rev3
[classic@framework alembic:main]$ PYTHONPATH=~/dev/sqlalchemy/lib/ python -m alembic current --verbose
INFO  [alembic.runtime.migration] Context impl PostgresqlImpl.
INFO  [alembic.runtime.migration] Will assume transactional DDL.
Current revision(s) for postgresql://scott:***@localhost/test:
Rev: 185d278c5837 (head)
Parent: 5f32c059b464
Path: /home/classic/dev/alembic/foo/versions/185d278c5837_rev2.py

    rev2
    
    Revision ID: 185d278c5837
    Revises: 5f32c059b464
    Create Date: 2024-05-29 12:22:04.837350


```

**Reply by zzzeek** [reply]
make sure you are on most recent alembic, this was not always possible

**Comment by lafrech** [comment]
Oh, boy. Error on my side.

My rev3 downgrade function was copied from the rev2 migration script and because I missed a few lines while find'n replacing, the downgrade function was basically the same as the rev2 one. It triggered the same FK error as when attempting the rev2 downgrade, so I assumed it was actually performing the rev2 downgrade as well.

Thanks for confirming the method was correct and sorry about the noise.
